### PR TITLE
Replace unmaintained actions-rs/toolchain action.

### LINKED ISSUE #1654: [Bug]: actions-rs/toolchain is archived
### What happened?

This repo relies on https://github.com/actions-rs/toolchain but that repo is archived.

The GitHub Actions test images have current versions of Rust tools preinstalled so this third party Action is probably no longer needed.  https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-rust+actions-rs&type=code

This out-of-date dependency causes several software supply chain warnings at the bottom right of
https://github.com/open-telemetry/opentelemetry-rust/actions/runs/8542099202

### API Version

Current

### SDK Version

Current

### What Exporters are you seeing the problem on?

_No response_

### Relevant log output

```shell
ALL
```

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #1654
Replacing `actions-rs/toolchain` action with more actively maintained and used `dtolnay/rust-toolchain`.  While `rustup` is available on Github actions runners, I feel it would be less of maintanance efffort with using an action. But happy to change if there are other thoughts.

Also, rust latest beta compiler/toolchain has more accurate redundant import warning - https://github.com/rust-lang/rust/pull/117772 - so fixed the code to remove redundant imports. Basically remove import for `TryInto`, `TryFrom` and `FromIterator` traits, as they are part of[ std::prelude](https://doc.rust-lang.org/stable/std/prelude/index.html).

## Changes

Please provide a brief description of the changes here.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by lalitb** [comment]
Ok there is error in workflow
```
Check failure on line 27 in .github/workflows/ci.yml

GitHub Actions
/ CI
Invalid workflow file
The workflow is not valid. .github/workflows/ci.yml (Line: 27, Col: 13): Unrecognized named-value: 'matrix'. Located at position 1 within expression: matrix.rust
```

Let me see how to fix this.

**Comment by lalitb** [comment]
The tests are failing with this toolchain action, moving it to draft to fix it.

File: stress/src/logs.rs [hunk 1]
```diff
@@ -1,39 +1,10 @@
 use opentelemetry_appender_tracing::layer;
 use opentelemetry_sdk::logs::{LogProcessor, LoggerProvider};
 use tracing::error;
-use tracing_subscriber::{prelude::*, Layer};
+use tracing_subscriber::prelude::*;
 
 mod throughput;
 
-struct NoopEventVisitor;
```
Comment (lalitb) [inline stress/src/logs.rs:8]: NoOpLogLayer and NoopEventVisitor are not used, so removing them,

**Review by cclauss — APPROVED** [review]
LGTM.

`.github/workflows/integration_tests.yml` is modified by this PR but that test is skipped.

File: stress/src/logs.rs (hunk 1 above)
Comment (hdost) [reply stress/src/logs.rs:8]: Do we not need them for the future?

Comment (lalitb) [reply stress/src/logs.rs:8]: I don't see them used for stress test.  We already have the benchmark test for them here - https://github.com/open-telemetry/opentelemetry-rust/blob/04a800c1b6a7f096c5040864592edb2e0e8fb9db/opentelemetry-appender-tracing/benches/logs.rs#L83

**Comment by lalitb** [comment]
> .github/workflows/integration_tests.yml is modified by this PR but that test is skipped.

Thanks, have forced run it now.

**Review by TommyCpp — COMMENTED** [review]
>  While rustup is available on Github actions runners, I feel it would be less of maintanance efffort with using an action

Agree with this. I prefer less change needed for CICD

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -24,11 +24,12 @@ jobs:
     - uses: actions/checkout@v4
       with:
         submodules: true
-    - uses: actions-rs/toolchain@v1
+    - uses: dtolnay/rust-toolchain@master
```
[the remark this answers is not in the document]
Comment (lalitb) [reply .github/workflows/ci.yml:27]: `stable` in L42 is the rustup toolchain specifier which can be `stable`, `nightly` or a specific version. `master` in L27 is the version of the action, and then we specify the toolchain separately as in L29 as it is a variable. 
For more details - https://github.com/dtolnay/rust-toolchain

[the remark this answers is not in the document]
Comment (TommyCpp) [reply .github/workflows/ci.yml:27]: 👍

**Review by TommyCpp — APPROVED** [review]
LGTM thanks!

**Comment by cclauss** [comment]
Oh...  Still two `lint` warnings at the bottom right of
https://github.com/open-telemetry/opentelemetry-rust/actions/runs/8559370429

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -38,11 +39,9 @@ jobs:
...
       with:
         submodules: true
-    - uses: actions-rs/toolchain@v1
+    - uses: dtolnay/rust-toolchain@stable
       with:
-        toolchain: stable
         components: rustfmt
-        profile: minimal
     - uses: arduino/setup-protoc@v3
     - uses: actions-rs/cargo@v1
```
Comment (cclauss) [inline .github/workflows/ci.yml:46]: Here...

**Comment by lalitb** [comment]
Looks like something has changed in the Github runner images, as existing PRs are failing with error:

```
thread 'build_tonic' panicked at opentelemetry-proto/tests/grpc_build.rs:116:9:
generated file has changed but it's a CI environment, please rerun this test locally and commit the changes
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: .github/workflows/ci.yml (hunk 3 above)
Comment (cclauss) [reply .github/workflows/ci.yml:46]: I would recommend that we bring back this use of `actions-rs/cargo@v1` to get the test working again and then after this is merged we can circle back to solve this one in a separate PR.

Comment (hdost) [reply .github/workflows/ci.yml:46]: @lalitb I'd agree

**Review by hdost — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 3 above)
Comment (lalitb) [reply .github/workflows/ci.yml:46]: @cclauss do you still see any issue here? The unrelated CI failure was fixed by #1658 .

**Comment by lalitb** [comment]
> Looks like something has changed in the GitHub runner images, as all existing PRs are failing with [error](https://github.com/open-telemetry/opentelemetry-rust/actions/runs/8573734574/job/23499043769?pr=1644):
> 
> ```
> thread 'build_tonic' panicked at opentelemetry-proto/tests/grpc_build.rs:116:9:
> generated file has changed but it's a CI environment, please rerun this test locally and commit the changes
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

All the issues are resolved. Should be good to merge.

**Pull request merged**